v1/util/cloud_events_util.cc

```diff
@@ -0,0 +1,110 @@
+#include "cloud_events_util.h"
+
+#include <google/protobuf/util/time_util.h>
+
+namespace cloudevents {
+namespace cloudevents_util {
+
+using ::io::cloudevents::v1::CloudEvent;
+using ::io::cloudevents::v1::CloudEvent_CloudEventAttribute;
+using ::google::protobuf::Timestamp;
+using ::google::protobuf::util::TimeUtil;
+
+constexpr char kErrCeInvalid[] = "Given Cloud Event is missing required attributes.";
+constexpr char kErrTimeInvalid[] = "Given time is invalid because it does not comply to RFC 3339.";
+constexpr char kErrAttrNotSet[] = "Given Cloud Event attribute is not set.";
+constexpr char kErrAttrNotHandled[] = "A Cloud Event type is not handled.";
+
+absl::Status CloudEventsUtil::IsValid(const CloudEvent& cloud_event) {
+  if (cloud_event.id().empty() ||
+      cloud_event.source().empty() ||
+      cloud_event.spec_version().empty() ||
+      cloud_event.type().empty()) {
+    return absl::InvalidArgumentError(kErrCeInvalid);
+  }
+  return absl::OkStatus();
+}
+
+absl::StatusOr<
+    absl::flat_hash_map<std::string, CloudEvent_CloudEventAttribute>>
+    CloudEventsUtil::GetMetadata(const CloudEvent& cloud_event) {
+  absl::Status is_valid = CloudEventsUtil::IsValid(cloud_event);
+  if (!is_valid.ok()) {
+    return is_valid;
+  }
+
+  // create absl::flat_hash_map from protobuf map of optional/ extensionattrs
+  absl::flat_hash_map<std::string, CloudEvent_CloudEventAttribute> attrs(
+    (cloud_event.attributes()).begin(),
+    cloud_event.attributes().end());
+
+  // insert required attrs
+  CloudEvent_CloudEventAttribute attr_val;
+  attr_val.set_ce_string(cloud_event.id());
+  attrs["id"] = attr_val;
+  attr_val.set_ce_string(cloud_event.source());
+  attrs["source"] = attr_val;
+  attr_val.set_ce_string(cloud_event.spec_version());
+  attrs["spec_version"] = attr_val;
+  attr_val.set_ce_string(cloud_event.type());
+  attrs["type"] = attr_val;
```

Comment:
Can you create a separate helper function e.g. GetEventAttributeFromCeString instead of reusing the same attr_val and resetting the ce_string